### PR TITLE
net: ipv6: Improve routing between interfaces

### PR DESCRIPTION
**Pull request by kderda**
This PR includes few fixes for routing IPv6 packet between interfaces. Every change is described in in commit's message.

### CODE REVIEW

**Review by jukkar — DISMISSED** [review]
LGTM

**Review by pdgendt — DISMISSED** [review]
Nit: fix the compliance issues

**Review by pdgendt — DISMISSED** [review]

**Comment by konradderda** [comment]
@rlubos wow, thanks a lot! I have cherry-picked the commit.

**Review by pdgendt — DISMISSED** [review]

**Review by jukkar — DISMISSED** [review]

File: tests/net/socket/misc/src/main.c [hunk 1]
```diff
@@ -310,23 +309,23 @@ void test_ipv4_so_bindtodevice(void)
...
 	zassert_true(sock_s >= 0, "socket open failed");
 
-	peer_addr.sin_family = AF_INET;
-	peer_addr.sin_port = htons(DST_PORT);
-	ret = zsock_inet_pton(AF_INET, TEST_PEER_IPV4_ADDR,
-			      &peer_addr.sin_addr);
-	zassert_equal(ret, 1, "inet_pton failed");
+	peer_addr_1.sin_family = AF_INET;
+	peer_addr_1.sin_port = htons(DST_PORT);
+	peer_addr_1.sin_addr = my_ipv4_addr1;
```
Comment (rlubos) [inline tests/net/socket/misc/src/main.c:314]: There's always something... It seems that userspace doesn't allow to reuse global variables for initialization... Weird it didn't manifest itself on native_sim, only on qemu platforms. TBH I'm not sure if there's a point digging into this, I'd fix it like this:
```patch
diff --git a/tests/net/socket/misc/src/main.c b/tests/net/socket/misc/src/main.c
index 1e22dada30e..a245c0f64e4 100644
--- a/tests/net/socket/misc/src/main.c
+++ b/tests/net/socket/misc/src/main.c
@@ -297,7 +297,16 @@ void test_ipv4_so_bindtodevice(void)
 {
        int sock_c;
        int sock_s;
-       struct sockaddr_in peer_addr_1, peer_addr_2;
+       struct sockaddr_in peer_addr_1 = {
+               .sin_family = AF_INET,
+               .sin_port = htons(DST_PORT),
+               .sin_addr = { { { 192, 0, 2, 1 } } },
+       };
+       struct sockaddr_in peer_addr_2 = {
+               .sin_family = AF_INET,
+               .sin_port = htons(DST_PORT),
+               .sin_addr = { { { 192, 0, 2, 2 } } },
+       };
        struct sockaddr_in bind_addr = {
                .sin_family = AF_INET,
                .sin_port = htons(DST_PORT),
@@ -309,13 +318,6 @@ void test_ipv4_so_bindtodevice(void)
        sock_s = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
        zassert_true(sock_s >= 0, "socket open failed");
 
-       peer_addr_1.sin_family = AF_INET;
-       peer_addr_1.sin_port = htons(DST_PORT);
-       peer_addr_1.sin_addr = my_ipv4_addr1;
-       peer_addr_2.sin_family = AF_INET;
-       peer_addr_2.sin_port = htons(DST_PORT);
-       peer_addr_2.sin_addr = my_ipv4_addr2;
-
        test_so_bindtodevice(sock_c, sock_s, (struct sockaddr *)&peer_addr_1,
                             (struct sockaddr *)&peer_addr_2, sizeof(peer_addr_1),
                             (struct sockaddr *)&bind_addr, sizeof(bind_addr));
@@ -325,7 +327,18 @@ void test_ipv6_so_bindtodevice(void)
 {
        int sock_c;
        int sock_s;
-       struct sockaddr_in6 peer_addr_1, peer_addr_2;
+       struct sockaddr_in6 peer_addr_1 = {
+               .sin6_family = AF_INET6,
+               .sin6_port = htons(DST_PORT),
+               .sin6_addr = { { { 0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+                                  0, 0, 0, 0, 0, 0, 0, 0x1 } } },
+       };
+       struct sockaddr_in6 peer_addr_2 = {
+               .sin6_family = AF_INET6,
+               .sin6_port = htons(DST_PORT),
+               .sin6_addr = { { { 0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+                                  0, 0, 0, 0, 0, 0, 0, 0x2 } } },
+       };
        struct sockaddr_in6 bind_addr = {
                .sin6_family = AF_INET6,
                .sin6_port = htons(DST_PORT),
@@ -337,13 +350,6 @@ void test_ipv6_so_bindtodevice(void)
        sock_s = zsock_socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
        zassert_true(sock_s >= 0, "socket open failed");
 
-       peer_addr_1.sin6_family = AF_INET6;
-       peer_addr_1.sin6_port = htons(DST_PORT);
-       peer_addr_1.sin6_addr = my_ipv6_addr1;
-       peer_addr_2.sin6_family = AF_INET6;
-       peer_addr_2.sin6_port = htons(DST_PORT);
-       peer_addr_2.sin6_addr = my_ipv6_addr2;
-
        test_so_bindtodevice(sock_c, sock_s, (struct sockaddr *)&peer_addr_1,
                             (struct sockaddr *)&peer_addr_2, sizeof(peer_addr_1),
                             (struct sockaddr *)&bind_addr, sizeof(bind_addr));
```

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**